File: Creational_Patterns/Factory_Pattern/python/Logistics.py
from abc import ABC, abstractmethod
from typing import override

class Logistics(ABC):
  @abstractmethod
  def send(self):
    pass

class Air(Logistics):
  @override
  def send(self):
    print("Sending By air")


class Road(Logistics):
  @override
  def send(self):
    print("Sending By Road")

class LogisticFactory:
  @staticmethod
  def get_logistics(mode):
    if mode.lower() == 'air':
      return Air()
      # logistics = Air()
      # logistics.send() # this is bad we have to have a service
    elif mode.lower() == 'road':
      return Road()
      # logistics = Road()
      # logistics.send() # this is bad we need to have a service
    return None
    
class LogisticsSerivce:
  def send(self, mode):
    logistics = LogisticFactory.get_logistics(mode)
    logistics.send()
  

if __name__ == '__main__':
  service = LogisticsSerivce()
  service.send("road")

File: Creational_Patterns/Factory_Pattern/java/Logistics.java
interface Logistics{
  public void send();
}

class Air implements Logistics{
  @Override
  public void send(){
    System.out.println("Sending via Air");
  }
}

class Road implements Logistics{
  @Override
  public void send(){
    System.out.println("Sending via Road");
  }
}

class LogisticsFactory{
  public static Logistics getLogistics(String mode){
    if(mode.equalsIgnoreCase("air")){
      return new Air();
    }else if(mode.equalsIgnoreCase("road")){
      return new Road();
    }else{
      return null;
    }
  }
}

class LogisticsService{
  public void send(String mode){
    Logistics logistics = LogisticsFactory.getLogistics(mode);
    logistics.send();
  }
}


class Main{
  public static void main(String[] args){
    LogisticsService logistics = new LogisticsService();
    logistics.send("air");
    logistics.send("road");
  }
}